Added in annotation vectors

nikita-smyrnov opened the pull request:
Added unit testing, documentation, citations, and functionality for annotation vectors in the matrix profile library as per specifications on the main website.

tylerwmarrs commented on matrixprofile/utils.py:
```diff
@@ -12,6 +12,272 @@
+        profile['av_cmp'] = av
+    elif av == "custom":
+        try:
+            profile['av'] = core.to_np_array(custom_av)
```
In this case, would profile['av_cmp'] be set to custom?

tylerwmarrs commented on matrixprofile/utils.py:
```diff
@@ -12,6 +12,272 @@
 from matrixprofile import core
 
 
+def apply_av(profile, av="default", custom_av=None):
```
This function has a problem with mutating the data structure before all of the validation passes. Here is a silly example you can play with to see the issue:

```python
profile = {'key': 1}

def change(p):
    p['key'] = 2

change(profile)
print(profile)
```
Note that in this silly example the key is changed similar to how you set the av and av_cmp keys. A better approach is to set these keys to variables and at the end mutate the profile data structure.

tylerwmarrs commented on matrixprofile/utils.py:
```diff
@@ -12,6 +12,272 @@
+        raise ValueError('make_complexity_av expects ts to be array-like')
+
+    if not core.is_one_dimensional(ts):
+        raise ValueError('make_complexity_av expects ts to be array-like')
```
The error message here should be related to dimensionality.

nikita-smyrnov commented:
I've fixed most of the issues, but there's one problem: for some reason, the python 2.7 build is failing due to the av_type key not being saved in the IO MPF format. I'm not particularly sure if Python 2.7 (or protobuf) handles strings differently than Python 3x. Perhaps there is something else missing that needs to be added to ensure backwards compatibility? Would appreciate some help in debugging this.

nikita-smyrnov commented:
I think it has something to do with basestring being the way to detect strings in Python 2x whereas str and bytes were used in Python 3x.  I added in some boilerplate code to fix the issue but it still doesn't seem to fully resolve the issue.

https://github.com/matrix-profile-foundation/matrixprofile/blob/2e30406a5c2ef652b3e6434c5ae956d2d9f99c39/matrixprofile/io/protobuf/protobuf_utils.py#L8

tylerwmarrs commented:
This seems to fix Python 2.7 compatibility. I also ran tests against 3.7.

[py2-protobuf-av.patch.txt](https://github.com/matrix-profile-foundation/matrixprofile/files/4483471/py2-protobuf-av.patch.txt)

nikita-smyrnov commented:
The patch you posted needed a slight change (check if string is empty before adding it to the MP data structure), but everything should be working now. As a summary of what has been fixed:
- Profile is no longer mutated before validation
- Original MP values are no longer altered; instead, a corrected matrix profile has been inserted as a separate key
- IO functionality has been added
- AV functionality has been to moved to mp.transform
- Error message was fixed

All the TravisCI and code coverage tests have also passed. Let me know if there's anything else you'd like to me handle for the AVs.

tylerwmarrs commented:
Thank you for all of the hard work on this feature. Everything is looking good. One thing that would be nice is to have a quick "tutorial/overview" on annotation vectors in the documentation. Not many people are aware of them and what they are used for. If you are up to it, adding an example in docs/examples as a jupyter notebook would be awesome.

Basically, include:
* high-level layman explanation
* a use case or two
* some sample code

nikita-smyrnov commented:
I've added in an example notebook where ECG time series data is analyzed. The first half contains a calibration signal motif that is better conserved that the heartbeat motif in the second half. The notebook then uses an AV to ignore the first half of the time series so that the important motif (the heartbeat signal) can be selected for.

I've also submitted a PR here to account for this additional dataset:
https://github.com/matrix-profile-foundation/mpf-datasets/pull/3

Also note that I've updated the motifs and visualize functions to account for this extra CMP/AV in the data structure. This way the user can select whether they want to use the MP or CMP when finding motifs, and visualize will generate a series of extra plots containing the updated CMP and AV. Documentation has also been updated.

Feel free to look at the example notebook for a visual example of how this occurs.

Let me know if there's anything else you'd like added in or edited for this feature.

tylerwmarrs commented:
This is good. A couple of modifications:

1. a CMP is not applicable to a PMP, however, you edited the top k motifs for PMP here: https://github.com/StateFarmIns/matrixprofile/blob/master/matrixprofile/algorithms/top_k_motifs.py#L16
2. The visualize function labels the AV as the CMP per your example (see screenshot). This may be fixed in your code already? I didn't see it there.

![image](https://user-images.githubusercontent.com/18449075/79611240-bf036100-80bf-11ea-872b-eb4bdd1cc69f.png)

3. The dataset pull request should have a "txt" extension as it is not formatted as "csv".

tylerwmarrs commented:
Unrelated to this code being merged, if you want to create a blog post on our MPF website, you can cross-post your example there. You will be acknowledged etc for it.

Here is the github repo:
https://github.com/matrix-profile-foundation/matrix-profile-foundation.github.io

Here is an example post:
https://github.com/matrix-profile-foundation/matrix-profile-foundation.github.io/blob/src/website/posts/nyc-taxi-analysis-anomalies-python.meta

The .meta file is just metadata around the post. The .ipynb file is the actual post content.

You will want to add an author image per the README file. :)

nikita-smyrnov commented:
1.  How would you want me to account for this since both PMP and MP motif functions use a general wrapper function to get called? I submitted a possible change, but I'm not sure if this is the best way to go about this. The only reason why I added it for the PMP motifs was to keep the original motif wrapper code, but I can see now why that would be confusing.

2. Fixed (was using an older version of code).

3. Fixed.

I'll definitely make sure to work on the blog post and will get that done once this pull request is completely finished and merged into the main repository.

tylerwmarrs commented:
I think your approach is good. There really isn't a nice way around it.
